File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:

        def maxCrossingSum(nums, left, mid, right):
            # Left half (max suffix sum)
            left_sum = float('-inf')
            total = 0
            for i in range(mid, left - 1, -1):
                total += nums[i]
                left_sum = max(left_sum, total)

            # Right half (max prefix sum)
            right_sum = float('-inf')
            total = 0
            for i in range(mid + 1, right + 1):
                total += nums[i]
                right_sum = max(right_sum, total)
            
            return left_sum + right_sum
        
        def maxSubArrayHelper(nums, left, right):
            if left == right:
                return nums[left]
            mid = (left + right) // 2
            left_sum = maxSubArrayHelper(nums, left, mid)
            right_sum = maxSubArrayHelper(nums, mid + 1, right)
            cross_sum = maxCrossingSum(nums, left, mid, right)

            return max(left_sum, right_sum, cross_sum)

        return maxSubArrayHelper(nums, 0, len(nums) - 1)

File: 0053-maximum-subarray/0053-maximum-subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        return maxSubArrayHelper(nums, 0, nums.length - 1);
    }

    private int maxSubArrayHelper(int[] nums, int left, int right) {
        if (left == right) {
            return nums[left];
        }
        int mid = (left + right) / 2;

        int leftSum = maxSubArrayHelper(nums, left, mid);
        int rightSum = maxSubArrayHelper(nums, mid + 1, right);
        int crossSum = maxCrossingSum(nums, left, mid, right);

        return Math.max(Math.max(leftSum, rightSum), crossSum);
    }

    private int maxCrossingSum(int[] nums, int left, int mid, int right) {
        // Left Half (max suffix sum)
        int left_sum = Integer.MIN_VALUE;
        int total = 0;
        for(int i = mid; i >= left; i--) {
            total += nums[i];
            left_sum = Math.max(total, left_sum);
        } 

        // Right Half (max prefix sum)
        int right_sum = Integer.MIN_VALUE;
        total = 0;
        for(int i = mid + 1; i < right + 1; i++) {
            total += nums[i];
            right_sum = Math.max(right_sum, total);
        }

        return (left_sum + right_sum);
    }
}